### PR TITLE
WIP: Add logic to support local dev environment

### PR DESCRIPTION
**Pull request by sigalsax**
We are pursuing an effort to build a local dev environment so that instead of pushing to a remote K8s cluster, developers can deploy the cluster locally with Docker desktop
- Add dev checks to block from unnecessary code from running
- Create separate bootstrap.env file for local development

TODO:
- [x] Add documentation for deploying Conjur for local development
- [x] Provide automated scripts for easy deployment with Secrets Provider

Feedback from: https://github.com/cyberark/kubernetes-conjur-deploy/pull/131/files

### CODE REVIEW

File: 1_prepare_conjur_namespace.sh [hunk 1]
```diff
@@ -62,7 +62,7 @@ create_cluster_role() {
 }
 
 create_cluster_role_binding() {
-  $cli delete --ignore-not-found clusterrolebinding conjur-authenticator-role-binding-$CONJUR_NAMESPACE_NAME
+  $cli delete --ignore-not-found rolebinding conjur-authenticator-role-binding-$CONJUR_NAMESPACE_NAME
```
Comment (orenbm) [inline 1_prepare_conjur_namespace.sh:65]: i see that you use a RoleBinding instead of a ClusterRoleBinding. Can you please split this into a separate commit and explain why you are performing this change? It will be interesting to know why RoleBinding is better.

Also - please change the method name while you're at it.

Comment (orenbm) [reply 1_prepare_conjur_namespace.sh:65]: oh i see now that it was always a rolebinding. that's odd. can you please just change the method name?

Also - i see that we have a RoleBinding and a ClusterRole. does that make sense? shouldn't it be a Role? (i know it's not part of your PR but you researched that area lately)

File: 0_check_dependencies.sh [hunk 2]
```diff
@@ -11,10 +11,18 @@ if [ ! is_minienv ]; then
   check_env_var "DOCKER_REGISTRY_PATH"
 fi
 
-if [ "${PLATFORM}" = "kubernetes" ] && [ ! is_minienv ]; then
+if [ "${PLATFORM}" = "kubernetes" ] && [ ! is_minienv ] ; then
```
Comment (orenbm) [inline 0_check_dependencies.sh:14]: you just added a space? is that intentional?

File: 0_check_dependencies.sh [hunk 3]
```diff
@@ -11,10 +11,18 @@ if [ ! is_minienv ]; then
...
   check_env_var "DOCKER_REGISTRY_URL"
 fi
 
+if [ "${DEV}" = "false" ]; then
+    check_env_var "DOCKER_REGISTRY_PATH"
+fi
+
+if [ "${PLATFORM}" = "kubernetes" ] && [ "${DEV}" = "false" ]; then
+      check_env_var "DOCKER_REGISTRY_PATH"
```
Comment (orenbm) [inline 0_check_dependencies.sh:23]: fix indentation

File: 0_check_dependencies.sh [hunk 4]
```diff
@@ -11,10 +11,18 @@ if [ ! is_minienv ]; then
...
+fi
```
Comment (orenbm) [inline 0_check_dependencies.sh:24]: this can be changed to 
```
if [ "${DEV}" = "false" ]; then
  check_env_var "DOCKER_REGISTRY_PATH"
  
  if [ "${PLATFORM}" = "kubernetes" ]; then
    check_env_var "DOCKER_REGISTRY_PATH"
  fi
fi
```

File: 2_prepare_docker_images.sh [hunk 5]
```diff
@@ -25,12 +25,17 @@ prepare_conjur_appliance_image() {
...
 
-  # Try to pull the image if we can
-  docker pull $CONJUR_APPLIANCE_IMAGE || true
-
-  docker tag $CONJUR_APPLIANCE_IMAGE $conjur_appliance_image
+  if [ "${DEV}" = "true" ]; then
+    docker pull registry.tld/conjur-appliance:5.0-stable
+    docker tag registry.tld/conjur-appliance:5.0-stable $conjur_appliance_image
+  else
+
```
Comment (orenbm) [inline 2_prepare_docker_images.sh:32]: please remove this blank line

File: 2_prepare_docker_images.sh [hunk 6]
```diff
@@ -25,12 +25,17 @@ prepare_conjur_appliance_image() {
 
   conjur_appliance_image=$(platform_image conjur-appliance)
 
-  # Try to pull the image if we can
-  docker pull $CONJUR_APPLIANCE_IMAGE || true
-
-  docker tag $CONJUR_APPLIANCE_IMAGE $conjur_appliance_image
+  if [ "${DEV}" = "true" ]; then
+    docker pull registry.tld/conjur-appliance:5.0-stable
```
Comment (orenbm) [inline 2_prepare_docker_images.sh:29]: do we want this value to be configurable?

for example, the `test.sh` script sets this to `registry.tld/conjur-appliance:5.0-stable` before it runs the `start` script. can we do something similar?

Also - i can see that this value is set, as it is being checked in `0_check_dependencies.sh`. This means that the person who ran `start` (and thus got to this script) has set `CONJUR_APPLIANCE_IMAGE` to some value. You now change this value to `registry.tld/conjur-appliance:5.0-stable` so it may cause an unexpected behaviour.

File: 4_deploy_conjur_followers.sh [hunk 7]
```diff
@@ -6,7 +6,9 @@ set -eo pipefail
 main() {
   set_namespace $CONJUR_NAMESPACE_NAME
 
-  docker_login
+  if [ "${DEV}" = "false" ]; then
+      docker_login
```
Comment (orenbm) [inline 4_deploy_conjur_followers.sh:10]: fix indentation

File: utils.sh [hunk 8]
```diff
@@ -6,9 +6,12 @@ DEPLOY_MASTER_CLUSTER="${DEPLOY_MASTER_CLUSTER:-false}"
 FOLLOWER_USE_VOLUMES="${FOLLOWER_USE_VOLUMES:-false}"
 
 MINI_ENV="${MINI_ENV:-false}"
+DEV="${DEV:-false}"
 
 if [[ "$MINI_ENV" == "true" ]]; then
   IMAGE_PULL_POLICY='Never'
+elif [[ "$DEV" == "true" ]]; then
+  IMAGE_PULL_POLICY='Never'
```
Comment (orenbm) [inline utils.sh:14]: can we merge these to one condition?

File: 1_prepare_conjur_namespace.sh (hunk 1 above)
Comment (sigalsax) [reply 1_prepare_conjur_namespace.sh:65]: Yeah I don't know clusterrolebinding was working in the past, maybe a clusterrole was once supplied instead of role. Anyways, I will update.

> Also - i see that we have a RoleBinding and a ClusterRole. does that make sense? shouldn't it be a Role? (i know it's not part of your PR but you researched that area lately
It does make sense because you can supply a ClusterRole and bind that role to a namespace for example, instead of having it be used across the cluster. We actually recommend this flow our [secrets provider docs as well](https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/Latest/en/Content/Integrations/Kubernetes_deployApplicationsConjur-k8s-Secrets.htm?Highlight=secrets%20provider#Step3ConfigureRBACpermissions)

File: 2_prepare_docker_images.sh (hunk 6 above)
Comment (sigalsax) [reply 2_prepare_docker_images.sh:29]: sure, I can add it to `dev-bootstrap.env` so it can be configurable
